src/layer.rs

```diff
@@ -175,17 +184,29 @@ impl LayerSet {
             Err(NamingError::Duplicate(new.to_string()))
         } else if self.get(old).is_none() {
             Err(NamingError::Missing(old.into()))
+        } else if new == DEFAULT_LAYER_NAME && self.layers[0].name != old {
+            Err(NamingError::ReservedName)
         } else {
-            match Name::new(new) {
-                Ok(name) => {
-                    if overwrite {
-                        self.layers.retain(|l| l.name != name)
-                    }
-                    self.get_mut(old).unwrap().name = name;
-                    Ok(())
-                }
-                Err(_) => Err(NamingError::Invalid(new.into())),
+            let name = Name::new(new)?;
+            if overwrite {
+                self.remove(&name);
+            }
+
+            // Dance around the borrow checker by using indices instead of references.
+            let layer_pos = self.layers.iter().position(|l| l.name.as_ref() == old).unwrap();
+
+            if layer_pos == 0 {
+                // Default layer: just change the name.
+                let layer = &mut self.layers[layer_pos];
+                layer.name = name;
+            } else {
+                // Non-default layer: change name and path.
+                let layer = &mut self.layers[layer_pos];
+                layer.path = crate::util::default_file_name_for_layer_name(&name).into();
+                layer.name = name;
```

Comment:
not a blocker but I find it more readable in general to use conditional blocks to only perform the conditional work, e.g:

```rust
let layer_pos = self.layers.iter().position(|l| l.name.as_ref() == old).unwrap();
if layer_pos != 0 {
    self.layers[layer_pos].path =
        crate::util::default_file_name_for_layer_name(&name).into();
}
self.layers[layer_pos].name = name;
```

Reply:
(I may have to undo this in the filenames PR but let's see.)